### PR TITLE
USWDS - File input: Enhance screen reader error callout

### LINKED ISSUE #6157: USWDS - Bug: File input error not announced to screen reader
### Describe the bug

When inputting an incorrect format in the file input, there is no indication to the screen reader that there is an error. While there is text present, this is not announced. This fails criteria [4.1.3, status message](https://www.w3.org/WAI/WCAG21/Understanding/status-messages.html). 

### Steps to reproduce the bug

1. Got to file input component preview
2. Open a screen reader
3. Go to file input for TXT and PDFs
4. Drag and drop an image file
5. Observe the text stating the error, but not error message is announced

### Expected Behavior

The user using the screen reader would hear that they have input the incorrect format.

### Related code

_No response_

### Screenshots

_No response_

### System setup

Chrome, PC
Screen reader used: JAWS

### Additional context

_No response_

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://designsystem.digital.gov/about/community/#community-conduct).
- [x] I checked the [current issues](https://github.com/uswds/uswds/issues?q=is%3Aissue+is%3Aopen+label%3A%22Type%3A+Bug%22) for duplicate bug reports.

### PR DESCRIPTION
**Pull request by mahoneycm**
# Summary

Fixed a bug that prevented screen readers from announcing the invalid file type error message. Now, screen readers will consistently read out the invalid file state.

## Breaking change

This is not a breaking change

## Related issue

Closes #6157 

## Related pull requests

[Change log →](https://github.com/uswds/uswds-site/pull/2933)

## Preview link

[File input specific variant →](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/cm-file-input-sr-error/?path=/story/components-form-inputs-file-input--specific)

## Problem statement

When entering an incorrect format in the file input, there's no indication to the screen reader that there's an error. Text present, but isn't not announced. This fails criteria [4.1.3, status message](https://www.w3.org/WAI/WCAG21/Understanding/status-messages.html).

## Solution

Bring error message into the `aria-label` to match SR instructions. This allows consistent callouts whenever the error message is present. It also avoids an issue where the focus doesn’t return to the browser if an SR user drags and drops a file, causing the error message alert to be skipped.

### Alternative solution considered

I attempted Mandy’s recommendation of adding `role="alert"` to the error message. It seemed to have worked at first but then continued to be ignored by screen readers.

## Major changes

- Error messages are now added to `aria-label`
- Visual error messages are now `aria-hidden` to avoid duplicate callouts
- Default error message now contains `Error:`
- Created `data-errorMessage` control in storybook preview

## Testing and review

1. Visit the [file input specific variant](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/cm-file-input-sr-error/iframe.html?args=&id=components-form-inputs-file-input--specific) page
2. Activate screen reader
3. Using your keyboard, navigate to the file input button
4. Activate the button and select a file to upload that is not a `.txt` or `.pdf` file.
5. Listen for error state to be read alongside instructions
6. Drag and drop another incorrect file into the input
7. Return focus to the component and confirm error state is read.
8. Activate file input once more
9. Use storybook controls to set a custom error message (`error_message` string control)
10. Confirm correct error message is displayed and ready by screen reader
11. Close out of the file browser without uploading a file
12. Confirm the error message has cleared.

> [!TIP]
It’s my understanding that screen reader users generally navigate via keyboard, so it’s important to test keyboard only navigation with this work.
> 

### Testing checklist

- [ ]  Error state is consistently read by screen readers
- [ ]  `aria-label` approach is logical and follows USWDS patterns
- [ ]  Code matches USWDS standards
- [ ] Accepted file message storybook control works as expected

### CODE REVIEW

**Review by amycole501 — APPROVED** [review]
Used JAWS to listen to the error and was able to hear it. Also, it read "error" which is great so thanks for that wording update.

The weird user interaction situation is that in order to acheive the error I had to drag a file from a file folder and not use the keyboard to select a file. Most SR users won't be dragging anything so it's possible the error won't be found by the majority of screen reader users.

But like I said above, the error IS now announced so I'm approving this.

**Review by amyleadem — CHANGES_REQUESTED** [review]
Thanks for your quick work on this, @mahoneycm. I was able to hear a the error message in VoiceOver in Firefox and Chrome, but did not hear it in Safari.

## Testing difficulties

I wanted to flag for other reviewers that it was a little difficult to test with VoiceOver because using "drag and drop" required interacting with both the file finder panel and the browser, and it took some effort to keep VO focus on the file input. To successfully get VO to read out the error message, I had to do the following steps:

1. Put focus on the file input with the keyboard by pressing the `tab` key
2. Once focused, I moved my cursor to the open file finder and immediately dragged an invalid file type into the component. 
    1. Pausing will cause VoiceOver to focus on the finder window and read out the file name. In this scenario, it does not move focus back to the component when the file is dragged and dropped.

## Thoughts
I am wondering if this interaction is an edge case since I would assume most screen reader users will not be using the mouse-driven drag and drop functionality. If the component is access via keyboard exclusively, the file panel disables invalid file types and prevents the user from focusing on them.  Because of this, no error message would be needed.

@amycole501 and @alex-hull is it possible this is a "passes with exceptions" issue instead?

File: packages/usa-file-input/src/index.js [hunk 1]
```diff
@@ -506,11 +507,18 @@ const preventInvalidFiles = (e, fileInputEl, instructions, dropTarget) => {
...
       removeOldPreviews(dropTarget, instructions);
       fileInputEl.value = ""; // eslint-disable-line no-param-reassign
       dropTarget.insertBefore(errorMessage, fileInputEl);
       errorMessage.textContent =
-        fileInputEl.dataset.errormessage || `This is not a valid file type.`;
+        fileInputEl.dataset.errormessage || defaultErrorMsg;
+
+      fileInputEl.setAttribute(
+        "aria-label",
+        `${defaultErrorMsg} ${DEFAULT_ARIA_LABEL_TEXT}`,
```
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-file-input/src/index.js:520]: Great catch! I'll work on adding this funcitonality

**Comment by amycole501** [comment]
I agree I think the drag and drop is an edge case since the error prevention is in place by using the keyboard functionality. Passes with exceptions allows us to also indicate that some browsers may not announce the error. I heard it in Chrome and JAWS and that covers the vast majority of users.

**Comment by mahoneycm** [comment]
@amyleadem  Re: Safari not reading the error message.

It seems like with Safari, it reads out the `aria-label` with error message _after_ just about all of the other instructions. This doesn't seem as helpful or alerting as we'd hope a message like this to be.

Safari does seem to consistently read out the "SR only status" every time the button is focused, but the other browsers do not. The other browsers only read out this hidden span while navigating through item by item.

I'm curious if it would be beneficial to update both sr only label and instructions in an attempt to cover all the bases? It seems like it wouldn't be too verbose since the browsers handle them differently but we would need to test with each browser and screen reader.

File: packages/usa-file-input/src/index.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-file-input/src/index.js]: I updated the logic to consider `data-errorMessage` and updated the Storybook controls to allow users to add the attribute via the `error_message` control!

Resolved in [39e5ffb](https://github.com/uswds/uswds/pull/6168/commits/39e5ffbe2e527630c68720f276e8e2def070df9e) and [d2eca47](https://github.com/uswds/uswds/pull/6168/commits/d2eca4784ee52675b8b52e772ec7d9aa46b3d9d0)

**Comment by mahoneycm** [comment]
I created #6197 to test updating the `sr-only` instructions instead of the `aria-label`. With this approach, the screen reader only reads the error message when screen reader users navigate by item. (For me, that's `ctrl + option + left/right`

When the focus returns to the button after an incorrect file is chosen, or if a user tabs to the file input, this error message will not be read.

There might be a befit of including the sr only update in our final fix but I wouldn't rely on this alone to callout the error state to the user.

**Comment by mahoneycm** [comment]
I also created #6200 to test adding `role="alert"` to the error message like originally suggested. This did not cause the error message to be read out consistently.

**Review by mejiaj — COMMENTED** [review]
Both solutions presented worked for me. So far this has been the most reliable solution in VoiceOver across browsers (Chromium & Safari). 

The enhancement to customize the error text is also nice, although users can't currently customize the instructions. Created issue #6206 for that.

### System info

- MacOS Sonoma 14.6.1
- Chromium 131
- Safari 17.6
- VoiceOver

### Tested

- [x] Dragging invalid file reads error message in VoiceOver (both Chrome & Safari)
- [x] Code quality

### Potential enhancements

Component doesn't tell you the type of file that's invalid. For single required file types this is OK. For multiple files, this could be confusing as you don't know what was rejected.

This could be a follow-up if out of scope.

File: packages/usa-file-input/src/index.js [hunk 2]
```diff
@@ -484,6 +484,7 @@ const preventInvalidFiles = (e, fileInputEl, instructions, dropTarget) => {
   if (acceptedFilesAttr) {
     const acceptedFiles = acceptedFilesAttr.split(",");
     const errorMessage = document.createElement("div");
+    errorMessage.setAttribute("aria-hidden", true);
```
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-file-input/src/index.js:487]: Correct! Keeping both may be a bit redundant

File: packages/usa-file-input/src/index.js [hunk 3]
```diff
@@ -506,11 +507,19 @@ const preventInvalidFiles = (e, fileInputEl, instructions, dropTarget) => {
 
     // If dragged files are not accepted, this removes them from the value of the input and creates and error state
     if (!allFilesAllowed) {
+      const errorMessageText =
+        fileInputEl.dataset.errormessage ||
+        "Error: This is not a valid file type.";
+
```
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-file-input/src/index.js]: I like the second recommendation. It's more consistent with the previous implementations.

Resolved in [6ea64ef](https://github.com/uswds/uswds/pull/6168/commits/6ea64ef0ea3fcda6db95abfee57d9cfd09733ec7)

File: packages/usa-file-input/src/usa-file-input.twig [hunk 4]
```diff
@@ -28,6 +28,7 @@
     {% if multiple -%} multiple{%- endif -%}
     {% if disabled_state == 'disabled' %} disabled{%- endif %}
     {% if disabled_state == "aria-disabled" -%} aria-disabled="true"{%- endif -%}
+    {% if error_message -%}data-errorMessage="{{ error_message }}"{%- endif -%}
```
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-file-input/src/usa-file-input.twig]: A couple things here...

## data-errorMessage

This is not a new data attribute. This already exists in the code and used to allow users to put custom "Accepted file message" when a user tries to upload a non-acceptable file type.

We provide guidance to the user to use `data-errorMessage` and changing this would be a breaking change for users:

> **Customizing the error message.** Add the data attribute data-errorMessage to usa-file-input to include a custom error message. [^1]

[^1]: [data-errorMessage guidance](https://arc.net/l/quote/qwrmgyzh)

## error_message control

I was incorrectly using a predefined content name `error_message` to set this value but opted to create a new `accepted_file_message` instead.

- `error_message` is used for the "helpful error message" that appears above the file input
- `accepted_file_message` will be used to set the optional `data-errorMessage` that appears inside the file input when an unacceptable file is selected
- I chose `accepted_file_message` because it matches the class of the displayed message: `usa-file-input__accepted-files-message`

I went ahead and named named both controls so they display cleanly:

image

Resolved in [655adc3](https://github.com/uswds/uswds/pull/6168/commits/655adc349b3f23b5c90a481f280e543f569a10bf)

File: packages/usa-file-input/src/index.js [hunk 5]
```diff
@@ -506,11 +507,19 @@ const preventInvalidFiles = (e, fileInputEl, instructions, dropTarget) => {
...
       removeOldPreviews(dropTarget, instructions);
       fileInputEl.value = ""; // eslint-disable-line no-param-reassign
       dropTarget.insertBefore(errorMessage, fileInputEl);
-      errorMessage.textContent =
-        fileInputEl.dataset.errormessage || `This is not a valid file type.`;
+      errorMessage.textContent = errorMessageText;
+
+      fileInputEl.setAttribute(
+        "aria-label",
+        `${errorMessageText} ${DEFAULT_ARIA_LABEL_TEXT}`,
```
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-file-input/src/index.js]: I like it!

Resolved in [abfb7ae](https://github.com/uswds/uswds/pull/6168/commits/abfb7ae5d8346cb8a18d2829377269ea9f1fda71)

File: packages/usa-file-input/src/usa-file-input.twig (hunk 4 above)
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-file-input/src/usa-file-input.twig]: Taking another look, that entire line seems redundant. We're performing a logic check before performing specific logic checks for each individual control. The only use I could see for it is to visually separate the optional attributes from the standard ones.

I opted to remove for now to reduce redundancy. Let me know if you'd prefer me to keep it and add `data-errorMessage`.

Removed in [8ee7a0f](https://github.com/uswds/uswds/pull/6168/commits/8ee7a0ff1b5b0504e2da1c37a20dcc86172aff65)

[the remark this answers is not in the document]
Comment (mejiaj) [reply packages/usa-file-input/src/usa-file-input.twig]: Thanks @mahoneycm, it makes sense to keep the existing data attribute.

We should make the storybook control label more readable, so I like you proposed change. It still feels slightly confusing to have it be labeled `data-errorMessage` and the control labeled `Accepted files message`, but not a blocker if it's consistent with rendered code.

**Review by mejiaj — CHANGES_REQUESTED** [review]
@mahoneycm a tricky issue, thanks for taking this on. The PR description and additional draft PR are helpful in understanding the issue. The feature itself for notifying users looks good.

Comments on the additional enhancements and template changes.

File: packages/usa-file-input/src/index.js [hunk 6]
```diff
@@ -484,6 +485,10 @@ const preventInvalidFiles = (e, fileInputEl, instructions, dropTarget) => {
   if (acceptedFilesAttr) {
     const acceptedFiles = acceptedFilesAttr.split(",");
     const errorMessage = document.createElement("div");
+    const userErrorText = fileInputEl.dataset.errormessage;
+    const errorMessageText = userErrorText || DEFAULT_ERROR_LABEL_TEXT;
```
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-file-input/src/index.js:489]: Created unit tests for default and custom error messages in [7714687](https://github.com/uswds/uswds/pull/6168/commits/7714687206b2c542bdb6c595b74a22e7d5aff595) and [cbe9486](https://github.com/uswds/uswds/pull/6168/commits/cbe9486aa73ec71a66b0e4cae83a3482809c04f9)

File: packages/usa-file-input/src/usa-file-input.stories.js [hunk 7]
```diff
@@ -16,6 +16,16 @@ export default {
...
+    error_message: {
+      name: "Error message",
+      control: { type: "text" },
+      defaultValue: "",
+    },
+    accepted_files_message: {
+      name: "Accepted files message",
+      control: { type: "text" },
+      defaultValue: "",
+    },
```
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-file-input/src/usa-file-input.stories.js:28]: There was a spacing issue with the attributes that caused it to collide with the `multiple` attribute. 

Resolved and working as of [345127d](https://github.com/uswds/uswds/pull/6168/commits/345127d89199c170082296908fedd88b8d4cf152)

File: packages/usa-file-input/src/usa-file-input.twig [hunk 8]
```diff
@@ -22,12 +22,11 @@
     class="usa-file-input"
     type="file"
     name="{{ id }}"
-    {% if hint or accept or multiple or disabled_state -%}
-    {% if hint -%}aria-describedby="{{ aria_hint }}"{%- endif -%}
-    {% if accept -%}accept="{{ accept }}"{%- endif -%}
+    {% if hint -%} aria-describedby="{{ aria_hint }}"{%- endif -%}
```
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-file-input/src/usa-file-input.twig]: Removing spaces causes all of the attributes to collide into one mangled attribute

image

I did remove the `-` which seemed to also trim whitespace before the next attribute.

Resolved in [345127d](https://github.com/uswds/uswds/pull/6168/commits/345127d89199c170082296908fedd88b8d4cf152)

File: packages/usa-file-input/src/usa-file-input.twig [hunk 9]
```diff
@@ -22,12 +22,11 @@
...
+    {% if accept -%} accept="{{ accept }}"{%- endif -%}
     {% if multiple -%} multiple{%- endif -%}
     {% if disabled_state == 'disabled' %} disabled{%- endif %}
     {% if disabled_state == "aria-disabled" -%} aria-disabled="true"{%- endif -%}
-    {% endif %}
+    {% if accepted_files_message -%} data-errorMessage="{{ accepted_files_message }}"{%- endif -%}
```
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-file-input/src/usa-file-input.twig]: Kept whitespace, removed `-` in [345127d](https://github.com/uswds/uswds/pull/6168/commits/345127d89199c170082296908fedd88b8d4cf152)

File: packages/usa-file-input/src/usa-file-input.twig [hunk 10]
```diff
@@ -22,12 +22,11 @@
     class="usa-file-input"
     type="file"
     name="{{ id }}"
-    {% if hint or accept or multiple or disabled_state -%}
-    {% if hint -%}aria-describedby="{{ aria_hint }}"{%- endif -%}
-    {% if accept -%}accept="{{ accept }}"{%- endif -%}
+    {% if hint -%} aria-describedby="{{ aria_hint }}"{%- endif -%}
+    {% if accept -%} accept="{{ accept }}"{%- endif -%}
```
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-file-input/src/usa-file-input.twig]: Kept whitespace, removed `-` in [345127d](https://github.com/uswds/uswds/pull/6168/commits/345127d89199c170082296908fedd88b8d4cf152)

File: packages/usa-file-input/src/usa-file-input.stories.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-file-input/src/usa-file-input.stories.js:28]: Also noting, this work does not introduce a new feature. The custom error message is pre-existing and [documented](https://arc.net/l/quote/xywqkveb) on site already

This change only affects some attributes to allow screen readers to read error message when focusing the file input button.

File: packages/usa-file-input/src/usa-file-input.twig (hunk 4 above)
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-file-input/src/usa-file-input.twig]: Yeah I agree. I had trouble naming the control because they're technically both error messages. I opted for the classname of the DOM element here. Let me know if you have another idea!

**Review by mlloydbixal — APPROVED** [review]
Tested and approved!

File: packages/usa-file-input/src/usa-file-input.twig (hunk 8 above)
[the remark this answers is not in the document]
Comment (mejiaj) [reply packages/usa-file-input/src/usa-file-input.twig]: @mahoneycm correct, that's why I noted we'll have to double-check. I can't remember _why_ these trim whitespace lines were added.

File: packages/usa-file-input/src/usa-file-input.twig [hunk 11]
```diff
@@ -22,12 +22,11 @@
...
-    {% if multiple -%} multiple{%- endif -%}
-    {% if disabled_state == 'disabled' %} disabled{%- endif %}
-    {% if disabled_state == "aria-disabled" -%} aria-disabled="true"{%- endif -%}
-    {% endif %}
+    {% if hint %} aria-describedby="{{ aria_hint }}"{% endif %}
+    {% if accept %} accept="{{ accept }}"{% endif %}
+    {% if multiple %} multiple{% endif %}
+    {% if disabled_state == 'disabled' %} disabled{% endif %}
+    {% if disabled_state == "aria-disabled" %} aria-disabled="true"{% endif %}
+    {% if accepted_files_message %} data-errorMessage="{{ accepted_files_message }}"{% endif %}
```
Comment (mejiaj) [inline packages/usa-file-input/src/usa-file-input.twig:30]: Are the extra spaces needed if we've removed the tags to strip whitespace?

```suggestion
    {% if hint %}aria-describedby="{{ aria_hint }}"{% endif %}
    {% if accept %}accept="{{ accept }}"{% endif %}
    {% if multiple %}multiple{% endif %}
    {% if disabled_state == 'disabled' %}disabled{% endif %}
    {% if disabled_state == "aria-disabled" %}aria-disabled="true"{% endif %}
    {% if accepted_files_message %}data-errorMessage="{{ accepted_files_message }}"{% endif %}
```

**Review by mejiaj — CHANGES_REQUESTED** [review]
Thanks for adding the tests, added few comments based on these updates!

File: packages/usa-file-input/src/index.js [hunk 12]
```diff
@@ -509,8 +514,11 @@ const preventInvalidFiles = (e, fileInputEl, instructions, dropTarget) => {
       removeOldPreviews(dropTarget, instructions);
       fileInputEl.value = ""; // eslint-disable-line no-param-reassign
       dropTarget.insertBefore(errorMessage, fileInputEl);
-      errorMessage.textContent =
-        fileInputEl.dataset.errormessage || `This is not a valid file type.`;
+      errorMessage.textContent = errorMessageText;
```
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-file-input/src/index.js:516]: Great catch! I think it was originally set this way to try to get the screen reader to call it out.

Resolved in [a020d98](https://github.com/uswds/uswds/pull/6168/commits/a020d98924ded5499262bcc865e670845c95a1fd)

File: packages/usa-file-input/src/test/file-input-accepts.spec.js [hunk 13]
```diff
@@ -134,6 +138,38 @@ tests.forEach(({ name, selector: containerSelector }) => {
           true,
         );
       });
+
+      it("should display an error message", () => {
+        // add to our elements FileList
+        inputEl.files = fileList;
+        const e = new Event("change");
+        inputEl.dispatchEvent(e);
```
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-file-input/src/test/file-input-accepts.spec.js]: I like this a lot! Since it uses the defined `inputEl` value, I made this an arrow function within in main tests rather than a function declaration outside of it. Let me know if that is the preferred approach!

Resolved in [31aae3c](https://github.com/uswds/uswds/pull/6168/commits/31aae3c383b5039f98e8867ca485fe99371239d8)

File: packages/usa-file-input/src/test/file-input-accepts.spec.js [hunk 14]
```diff
@@ -134,6 +138,38 @@ tests.forEach(({ name, selector: containerSelector }) => {
...
+
+        // Error message appended to DOM after change event.
+        errorMessage = body.querySelector(
+          ".usa-file-input__accepted-files-message",
```
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-file-input/src/test/file-input-accepts.spec.js]: Resolved in [5feb026](https://github.com/uswds/uswds/pull/6168/commits/5feb0263d435f952811577ad3d7ddac91077f9d1)

File: packages/usa-file-input/src/test/file-input-accepts.spec.js [hunk 15]
```diff
@@ -134,6 +138,38 @@ tests.forEach(({ name, selector: containerSelector }) => {
           true,
         );
       });
+
+      it("should display an error message", () => {
```
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-file-input/src/test/file-input-accepts.spec.js]: With the current approach, each test is scoped to a single task. The `mock file should not be allowed` test looks for the invalid file class to confirm while this check tests the error messages specifically.

Alternatively, we could make a more encompassing, "error state is triggered" test that searches for 

- Invalid file classes
- Error state classes
- Error messages

Do you have a preference?

[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-file-input/src/test/file-input-accepts.spec.js]: Test description updated in [6bd5173](https://github.com/uswds/uswds/pull/6168/commits/6bd517353662efb36de340709ef43614ab90cae2)

File: packages/usa-file-input/src/test/file-input-accepts.spec.js [hunk 16]
```diff
@@ -134,6 +138,38 @@ tests.forEach(({ name, selector: containerSelector }) => {
...
+        errorMessage = body.querySelector(
+          ".usa-file-input__accepted-files-message",
+        );
+        ariaLabel = inputEl.getAttribute("aria-label");
+
+        assert.strictEqual(errorMessage.textContent, defaultErrorMessage);
+        assert.strictEqual(ariaLabel.startsWith(defaultErrorMessage), true);
+      });
+
+      it("should use custom error message", () => {
```
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-file-input/src/test/file-input-accepts.spec.js]: Resolved in [b37225d](https://github.com/uswds/uswds/pull/6168/commits/b37225dfd732ad512eb1af2625010fb5bd6a45c3)

File: packages/usa-file-input/src/usa-file-input.twig (hunk 11 above)
Comment (mahoneycm) [reply packages/usa-file-input/src/usa-file-input.twig]: Thank you for flagging this! I missed that I could remove the spaces now that I've removed the strip whitespace flag!

Resolved in [39e16e6](https://github.com/uswds/uswds/pull/6168/commits/39e16e6942b8e761aeadd3ae05baa56ea0991068)

Comment (mejiaj) [reply packages/usa-file-input/src/usa-file-input.twig]: Thank you! 

Ran the build HTML script [`npm run build:html`]. No regressions found in templates that we use on site.

**Review by mejiaj — APPROVED** [review]

**Comment by mahoneycm** [comment]
@mlloydbixal for this component, I don't think the delay is necessary because the user won't be inputting text the way it does for character count or input mask. File input is likely a singular input of the file(s), which should let us immediately add the error message. 

When testing, do you see any potential issues with how the message is read in this PR?

**Review by mlloydbixal — DISMISSED** [review]
On testing step number 4 I was unable to force an error. After triggering the select file window and only providing incorrect file types the screen reader lets me know there is no selection available. This is the behavior I would expect. In other steps where an error was possible, the screen reader and error message were triggered correctly.

**Review by alex-hull — DISMISSED** [review]
Tested with both JAWS and NVDA and they sound good!

File: packages/usa-file-input/src/usa-file-input.stories.js (hunk 7 above)
Comment (amyleadem) [inline packages/usa-file-input/src/usa-file-input.stories.js:28]: **issue:** I found the error controls in these stories to be confusing. I was trying to customize the "Error message" field thinking that it would represent `data-errorMessage`. I didn't expect that data attribute to be represented by a control named "Accepted files message". Can you add some clarity to the control names here? 

If we could also rename the "error" control to something more descriptive ("Form group error styles"?), that would be helpful as well. 

Feel free to edit this, but here is an idea:

```suggestion
    error_hint_text: {
      name: "Error text - hint",
      control: { type: "text" },
      defaultValue: "",
    },
    invalid_file_text: {
      name: "Error text - Invalid file type (data-errorMessage)",
      control: { type: "text" },
      defaultValue: "",
    },
```

**Review by amyleadem — CHANGES_REQUESTED** [review]
Thanks for this @mahoneycm. I am able to hear the invalid file type readout in most browsers. I have a couple of suggestions in the comments, but I think the code is looking good. 

Some notes:

- I was only able to trigger the invalid file type message by dragging and dropping. My OS prevented me from selecting an invalid file type from the file finder (Please let me know if there is a way around this)
- I also still do not hear the readout in Safari when dragging an invalid file type into the component. However, I do not consider this to be a blocker for the following reasons:
  - @mejiaj  and @mahoneycm  report being able to hear the error message in Safari/VoiceOver. After digging a bit deeper, I realized that I don’t hear the `aria-label` announced for any of the file inputs that also have an `aria-describedby` attribute, whether they are in an error state or not. This is happening on `develop` too, so I am counting this as a quirk on my local machine and looking into what is going on with local configuration. 
  - Combining drag and drop actions with a screen reader is likely an edge case for users. I don't think it is necessary to get 100% coverage to count this as a win. 

I was able to complete the following tests:

- [ ]  Confirm that the VoiceOver readout includes the error message when an invalid file type is added
    - [x]  Chromium
    - [ ]  Safari - I do not hear the error message in the Safari/VoiceOver readout. I do not consider this be a blocker because other testers on the team do hear it (more details in the note above).
    - [x]  Firefox
- [x]  Confirm custom error message (added via `data-errorMessage`) updates as expected for visual displays and screen reader readouts
- [ ]  Confirm the Storybook controls work as expected
    - The controls work as expected, but the control names were confusing. Recommend we update it to improve interaction in the future (See more details in [this comment](https://github.com/uswds/uswds/pull/6168#discussion_r1879066829)).
- [x]  Confirm unit tests are present and work as expected
- [ ]  Confirm the PR description is up-to-date and accurate
    - Recommend we update the summary to “Fixed a bug that prevented screen readers from announcing the invalid file type error message.”
    - Note (non-blocking): I was not able to follow the testing instructions as listed because I was only able to trigger the invalid file type message by dragging and dropping (The test says I should be able to select a file by activating the button). My OS prevented me from selecting an invalid file type from the file finder.

File: packages/usa-file-input/src/test/file-input-accepts.spec.js [hunk 17]
```diff
@@ -126,14 +136,43 @@ tests.forEach(({ name, selector: containerSelector }) => {
...
-        const e = new Event("change");
-        inputEl.dispatchEvent(e);
+        addFiles();
         assert.strictEqual(
           dropZone.classList.contains(INVALID_FILE_CLASS),
           true,
         );
       });
+
+      it("should provide a default error state", () => {
```
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-file-input/src/test/file-input-accepts.spec.js]: Resolved in [f7a06e4](https://github.com/uswds/uswds/pull/6168/commits/f7a06e4e80e6dff1e89e6e4091b3dbef33ccd4fd)

File: packages/usa-file-input/src/test/file-input-accepts.spec.js [hunk 18]
```diff
@@ -126,14 +136,43 @@ tests.forEach(({ name, selector: containerSelector }) => {
...
+        ariaLabel = inputEl.getAttribute("aria-label");
+
+        assert.strictEqual(
+          visibleErrorMessage.textContent,
+          defaultErrorMessage,
+        );
+        assert.strictEqual(ariaLabel.startsWith(defaultErrorMessage), true);
+      });
+
+      it("should allow a custom error message", () => {
```
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-file-input/src/test/file-input-accepts.spec.js]: Resolved in [f7a06e4](https://github.com/uswds/uswds/pull/6168/commits/f7a06e4e80e6dff1e89e6e4091b3dbef33ccd4fd)

**Review by amyleadem — APPROVED** [review]
LGTM!

I still do not hear the aria-label readout in VoiceOver in Safari, but this is a pre-existing issue with `develop` and one that seems to be specific to my machine. I am not considering this issue to be a blocker for this PR because there is a distinct improvement in Firefox and Chromium browsers. 

- [ ]  Confirm that the VoiceOver readout includes the error message when an invalid file type is added
    - [x]  Chromium
    - [ ]  Safari - I do not hear the error message in the Safari/VoiceOver readout. I do not consider this be a blocker because other testers on the team do hear it. I suspect it is a local issue for me. 
    - [x]  Firefox
- [x]  Confirm custom error message (added via `data-errorMessage`) updates as expected for visual displays and screen reader readouts
- [x]  Confirm the Storybook controls work as expected
- [x]  Confirm unit tests are present and work as expected
- [x]  Confirm the PR description is up-to-date and accurate

> [!note]
> Non-blocking issue with testing instructions: I am still not able to select an invalid file format with keyboard only, even after following specific instructions from @mahoneycm. I suspect this might be related to an OS update that I need to do. It ultimately does not impact this review because I was able to trigger the error message via drag and drop.

**Review by thisisdano — APPROVED** [review]
